packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart

```diff
@@ -0,0 +1,220 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:amplify_api/src/decorators/web_socket_auth_utils.dart';
+import 'package:amplify_api/src/graphql/web_socket/blocs/web_socket_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/state/web_socket_state.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/web_socket_types.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/ws_subscriptions_event.dart';
+import 'package:amplify_api/src/graphql/web_socket/web_socket_message_stream_transformer.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:async/async.dart';
+import 'package:flutter/material.dart';
+import 'package:stream_transform/stream_transform.dart';
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+/// {@template api.web_socket_service}
+/// Internal stateless [WebSocketService] abstract.
+/// {@endtemplate}
+abstract class WebSocketService implements Closeable {
+  /// Used to initialize a [WebSocketChannel] connection and return transformed
+  /// messages as [WebSocketEvent]
+  Stream<WebSocketEvent> init(
+    WebSocketState state,
+  );
+
+  /// Subscribe a [GraphQLRequest] inside the provided [WsSubscriptionBloc] to a open [WebSocketChannel]
+  Future<void> register(
+    ConnectedState state,
+    GraphQLRequest<Object?> request,
+  );
+
+  /// Unsubscribe a [GraphQLRequest] inside the provided [WsSubscriptionBloc]
+  Future<void> unsubscribe(
+    String subscriptionId,
+  );
+
+  /// Reconnect a [WebSocketChannel]
+  Future<void> reconnect();
+}
+
+/// {@template api.web_socket_service}
```

Comment:
Template strings are typically of the form `<package name>.<class name>` since these template strings need to be _globally_ unique across all packages. Sometimes they'll be `<package name>.<some namespace>.<class name>` if there are multiple classes with the same name in the package.

```suggestion
/// {@template amplify_api.web_socket_service}
```

Reply:
Thanks for the info. Updated other instances too 👍